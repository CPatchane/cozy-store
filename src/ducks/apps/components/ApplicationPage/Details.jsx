import React, { Component } from 'react'
import { withRouter } from 'react-router-dom'

import { translate } from 'cozy-ui/react/I18n'

import { Button } from 'cozy-ui/react/Button'
import Icon from 'cozy-ui/react/Icon'
import Toggle from 'cozy-ui/react/Toggle'

import ReactMarkdownWrapper from 'ducks/components/ReactMarkdownWrapper'
import { getContext, REGISTRY_CHANNELS } from 'ducks/apps'
import getChannel from 'lib/getChannelFromSource'

const isValidUrl = url => {
  if (!url) return null
  return url.match(
    /^(?:http(s)?:\/\/)?[\w.-]+(?:\.[\w.-]+)+[\w\-._~:/?#[\]@!$&'()*+,;=.]+$/
  )
}

const isLessButtonNeeded = (text = '') => {
  // Temporary disbable the less behaviour
  return false
  // const screenWidth = window.innerWidth || 0
  // const textNewLinesCount = text.split(/\r\n|\r|\n/).length
  // if (screenWidth <= 768) {
  //   return text.length > 350 || textNewLinesCount > 7
  // } else if (screenWidth <= 1024) {
  //   return text.length > 220 || textNewLinesCount > 7
  // } else if (screenWidth <= 1440) {
  //   return text.length > 350 || textNewLinesCount > 7
  // } else if (screenWidth <= 2400) {
  //   return text.length > 500 || textNewLinesCount > 7
  // } else {
  //   return false
  // }
}

export class Details extends Component {
  constructor(props) {
    super(props)
    const { changes, description, source } = this.props
    const appChannel = getChannel(source)
    this.state = {
      lessDescription: isLessButtonNeeded(description),
      lessChanges: isLessButtonNeeded(changes),
      displayBetaChannel: appChannel === REGISTRY_CHANNELS.BETA,
      displayDevChannel: appChannel=== REGISTRY_CHANNELS.DEV
    }

    this.toggleChannels = this.toggleChannels.bind(this)
    this.onUpdateChannel = this.onUpdateChannel.bind(this)
  }

  onShowPermissions () {
    const { app, parent } = this.props
    this.props.history.push(`/${parent}/${app.slug}/permissions`)
  }

  toggleDisplayMore (type) {
    this.setState({ [`less${type}`]: false })
  }

  toggleChannels() {
    const { source } = this.props
    if (!getChannel(source)) console.warn('This application don\'t use the registry')
    if (!this.state.displayBetaChannel) {
      this.setState(state => ({ displayBetaChannel: true }))
    }
    if (!this.state.displayDevChannel) {
      getContext()
      .then(context => {
        if (context && context.attributes && context.attributes.debug) {
          this.setState(state => ({ displayDevChannel: true }))
        }
      })
    }
  }

  onUpdateChannel(checked, channel) {
    const { slug, history, parent } = this.props
    const targetChannel = checked ? channel : REGISTRY_CHANNELS.STABLE
    history.push(`/${parent}/${slug}/channel/${targetChannel}`)
  }

  render() {
    const {
      t,
      slug,
      source,
      description,
      changes,
      categories,
      langs,
      mobileApps,
      developer,
      version
    } = this.props
    const {
      lessDescription,
      lessChanges,
      displayBetaChannel,
      displayDevChannel
    } = this.state
    const appChannel = getChannel(source)
    const isBeta = appChannel === REGISTRY_CHANNELS.BETA
    const isDev = appChannel === REGISTRY_CHANNELS.DEV
    const langsInfos = langs && langs.map(l => t(`app_langs.${l}`))
    const categoriesInfos =
      categories &&
      !!categories.length &&
      categories.map(c => t(`app_categories.${c}`))
    const developerName =
      developer && developer.name === 'Cozy'
        ? 'Cozy Cloud Inc.'
        : developer.name
    const shortVersion = version && version.match(/^(\d+\.\d+\.\d+)-.*$/)
    const displayedVersion =
      (shortVersion && shortVersion.length && shortVersion[1]) || version
    return (
      <div className="sto-app-details">
        <div className="sto-app-descriptions">
          {description && (
            <div className="sto-app-description">
              <h3>{t('app_page.description')}</h3>
              <ReactMarkdownWrapper
                source={description}
                className={lessDescription ? 'sto-app-description--less' : ''}
                parseEmoji
              />
              {lessDescription && (
                <button
                  type="button"
                  className="sto-details-display-more"
                  onClick={() => this.toggleDisplayMore('Description')}
                >
                  {t('app_page.more')}
                </button>
              )}
            </div>
          )}
          {changes && (
            <div className="sto-app-changes">
              <h3>{t('app_page.changes')}</h3>
              <ReactMarkdownWrapper
                source={changes}
                className={lessChanges ? 'sto-app-changes--less' : ''}
                parseEmoji
              />
              {lessChanges && (
                <button
                  type="button"
                  className="sto-details-display-more"
                  onClick={() => this.toggleDisplayMore('Changes')}
                >
                  {t('app_page.more')}
                </button>
              )}
            </div>
          )}
        </div>
        <div className="sto-app-additional-details">
          <h3>{t('app_page.infos.title')}</h3>
          <div className="sto-app-info">
            <div className="sto-app-info-header">
              {t('app_page.infos.categories')}
            </div>
            <div className="sto-app-info-content">
              {categoriesInfos &&
                categoriesInfos.join(t('app_categories.list_separator'))}
            </div>
          </div>
          <div className="sto-app-info">
            <div className="sto-app-info-header">
              {t('app_page.infos.version.title')}
            </div>
            <div className="sto-app-info-content">
              <span onDoubleClick={this.toggleChannels}>
                {displayedVersion || t('app_page.infos.version.unknown')}
              </span>
            </div>
          </div>
          {(displayBetaChannel || displayDevChannel) &&
            <div className="sto-app-info">
              <div className="sto-app-info-header">
                {t('app_page.infos.beta')}
              </div>
              <div className="sto-app-info-content">
                <Toggle
                  id={`sto-app-${slug}-beta-toggle`}
                  checked={isBeta}
                  onToggle={(e) => this.onUpdateChannel(e, REGISTRY_CHANNELS.BETA)}
                />
              </div>
            </div>
          }
          {displayDevChannel &&
            <div className="sto-app-info">
              <div className="sto-app-info-header">
                {t('app_page.infos.dev')}
              </div>
              <div className="sto-app-info-content">
                <Toggle
                  id={`sto-app-${slug}-dev-toggle`}
                  checked={isDev}
                  onToggle={(e) => this.onUpdateChannel(e, REGISTRY_CHANNELS.DEV)}
                />
              </div>
            </div>
          }
          {langsInfos && (
            <div className="sto-app-info">
              <div className="sto-app-info-header">
                {t('app_page.infos.langs')}
              </div>
              <div className="sto-app-info-content">
                {langsInfos.join(t('app_langs.list_separator'))}
              </div>
            </div>
          )}
          {mobileApps &&
            !!mobileApps.length && (
              <div className="sto-app-info">
                <div className="sto-app-info-header">
                  {t('app_page.infos.mobile_app')}
                </div>
                <div className="sto-app-info-content sto-app-info-content--mobile-apps">
                  {mobileApps.map(a => {
                    const icon = require(`assets/icons/platforms/icon-${
                      a.type
                    }.svg`)
                    return (
                      <a
                        className="sto-app-info-content-icon"
                        href={isValidUrl(a.url) ? a.url : null}
                        target="_blank"
                        key={a.type}
                      >
                        <Icon icon={icon.default} width="24px" height="24px" />
                      </a>
                    )
                  })}
                </div>
              </div>
            )}
          {developerName && (
            <div>
              <h3>{t('app_page.developer_infos')}</h3>
              <div className="sto-app-developer-infos">
                <span>{developerName}</span>
                {isValidUrl(developer.url) && (
                  <a
                    href={developer.url}
                    className="sto-app-developer-link"
                    target="_blank"
                  >
                    {developer.url}
                  </a>
                )}
              </div>
            </div>
          )}
          <div>
            <Button
              label={t('app_page.permissions.button.label')}
              onClick={() => this.onShowPermissions()}
              subtle
            />
          </div>
        </div>
      </div>
    )
  }
}

export default withRouter(translate()(Details))
