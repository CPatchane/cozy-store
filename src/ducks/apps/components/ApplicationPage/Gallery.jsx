import React from 'react'

import { translate } from 'cozy-ui/react/I18n'

export const Gallery = ({t, slug, images}) => {
  if (!images) return null
  const imagesParts = images.reduce((acc, image, index) => {
    if (index < 2) {
      if (!acc.big) acc.big = []
      acc.big.push(image)
    } else if (index < 5) {
      if (!acc.small) acc.small = []
      acc.small.push(image)
    }
    return acc
  }, {})
  return (
    <div className='sto-app-images'>
      {imagesParts.big && !!imagesParts.big.length &&
        <div className='sto-app-big-images'>
          {imagesParts.big.map((image, index) =>
            <img className='sto-app-image' src={image} alt={`${slug}-image-${index + 1}`} />
          )}
        </div>
      }
      {imagesParts.small && !!imagesParts.small.length &&
        <div
          className={`sto-app-small-images${imagesParts.small.length > 2 ? ' --space-around' : ''}`}
        >
          {imagesParts.small.map((image, index) =>
            <img className='sto-app-small-image' src={image} alt={`${slug}-image-${index + 3}`} />
          )}
        </div>
      }
    </div>
  )
}

export default translate()(Gallery)
