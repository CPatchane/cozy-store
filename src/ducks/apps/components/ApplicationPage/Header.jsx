import React from 'react'
import { Link } from 'react-router-dom'

import Icon from 'cozy-ui/react/Icon'
import Button from 'cozy-ui/react/Button'
import { translate } from 'cozy-ui/react/I18n'

import cozySmileIcon from 'assets/icons/icon-cozy-smile.svg'
import defaultAppIcon from 'assets/icons/icon-cube.svg'

import { APP_TYPE } from 'ducks/apps'

export const Header = ({
  t,
  icon,
  slug,
  namePrefix,
  name,
  type,
  description,
  installed,
  installedAppLink,
  parent
}) => {
  const openApp = link => {
    window.location.assign(link)
  }
  const isKonnector = type === APP_TYPE.KONNECTOR
  return (
    <div className='sto-app-header'>
      <div className='sto-app-header-icon'>
        {icon ? (
          <img className='sto-app-icon' src={icon} alt={`${slug}-icon`} />
        ) : (
          <Icon
            className='sto-app-icon--default blurry'
            icon={defaultAppIcon}
            height='88px'
            width='88px'
            color='#95999D'
          />
        )}
      </div>
      <div className='sto-app-header-content'>
        <h2 className='sto-app-header-title'>
          {namePrefix ? `${namePrefix} ${name}` : name}
        </h2>
        <p>{description}</p>
        {installed ? (
          <div>
            <Button
              onClick={() => openApp(installedAppLink)}
              className="c-btn"
              icon='openwith'
              label={isKonnector
                ? t('app_page.konnector.open')
                : t('app_page.webapp.open')
              }
            />
            <Link
              to={`/${parent}/${slug}/manage`}
              className='c-btn c-btn--danger-outline sto-app-header-uninstall-button'
            >
              <span>{isKonnector
                ? t('app_page.konnector.uninstall')
                : t('app_page.webapp.uninstall')
              }</span>
            </Link>
          </div>
        ) : (
          <Link
            to={`/${parent}/${slug}/manage`}
            className='c-btn c-btn--regular'
          >
            <span>
              <Icon
                icon={cozySmileIcon}
                color='#FFFFFF'
                width='16px'
                height='16px'
                className='sto-app-icon--button'
              />{' '}
              {isKonnector
                ? t('app_page.konnector.install')
                : t('app_page.webapp.install')
              }
            </span>
          </Link>
        )}
      </div>
    </div>
  )
}

export default translate()(Header)
