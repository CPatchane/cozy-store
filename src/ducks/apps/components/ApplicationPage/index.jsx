/* global cozy */
import React, { Component } from 'react'
import { Link } from 'react-router-dom'

import { translate } from 'cozy-ui/react/I18n'
import Spinner from 'cozy-ui/react/Spinner'
import withBreakpoints from 'cozy-ui/react/helpers/withBreakpoints'
import Button from 'cozy-ui/react/Button'
import FocusTrap from 'focus-trap-react'

import Header from './Header'
import Gallery from './Gallery'
import Details from './Details'

import {
  preventBackgroundScroll,
  unpreventBackgroundScroll
} from 'lib/scrollHelpers.js'

const MOBILE_PLATFORMS = ['ios', 'android']
const isMobilePlatform = name => MOBILE_PLATFORMS.includes(name)

const { BarCenter } = cozy.bar

export class ApplicationPage extends Component {
  constructor(props) {
    super(props)
    this.state = { displayBarIcon: false, activeTrap: true }
  }

  componentDidMount() {
    preventBackgroundScroll()
    const { mainPageRef } = this.props
    if (mainPageRef && mainPageRef.current) {
      mainPageRef.current.addEventListener('scroll', this.handleScroll, {
        passive: true
      })
    }
  }

  componentWillUnmount() {
    unpreventBackgroundScroll()
    const { mainPageRef } = this.props
    if (mainPageRef && mainPageRef.current) {
      mainPageRef.current.removeEventListener('scroll', this.handleScroll)
    }
  }

  mountTrap = () => {
    this.setState({ activeTrap: true })
  }

  unmountTrap = () => {
    this.setState({ activeTrap: false })
  }

  handleScroll = () => {
    if (
      !this.state.displayBarIcon &&
      this.props.mainPageRef.current.scrollTop > 100
    ) {
      this.setState(() => ({ displayBarIcon: true }))
    } else if (
      this.state.displayBarIcon &&
      this.props.mainPageRef.current.scrollTop < 100
    ) {
      this.setState(() => ({ displayBarIcon: false }))
    }
  }

  render() {
    const {
      t,
      parent,
      app,
      isFetching,
      fetchError,
      breakpoints = {},
      pauseFocusTrap
    } = this.props
    const { displayBarIcon } = this.state
    const { isMobile } = breakpoints
    if (isFetching) {
      return (
        <div className="sto-app">
          <Spinner size="xxlarge" loadingType="appsFetching" middle />
        </div>
      )
    }
    if (fetchError) {
      return (
        <p className="u-error">
          {t('app_modal.install.message.version_error', {
            message: fetchError.message
          })}
        </p>
      )
    }
    const { icon, slug, iconToLoad } = app
    const appName = t(`apps.${app.slug}.name`, { _: app.name })
    const namePrefix = t(`apps.${app.slug}.name_prefix`, {
      _: app.name_prefix || ''
    })
    const appShortDesc = t(`apps.${app.slug}.short_description`, {
      _: app.short_description || ''
    })
    const appLongDesc = t(`apps.${app.slug}.long_description`, {
      _: app.long_description || ''
    })
    const appChanges = t(`apps.${app.slug}.changes`, { _: '' })
    const mobileApps =
      app.platforms &&
      !!app.platforms.length &&
      app.platforms.reduce((mobilePlatforms, platformInfos) => {
        // force to be lower case
        platformInfos.type = platformInfos.type.toLowerCase()
        if (isMobilePlatform(platformInfos.type)) {
          mobilePlatforms.push(platformInfos)
        }
        return mobilePlatforms
      }, [])
    return (
      <FocusTrap
        focusTrapOptions={{
          onDeactivate: this.unmountTrap,
          clickOutsideDeactivates: true
        }}
        paused={pauseFocusTrap}
      >
        <div className="sto-modal-page-app">
          {isMobile &&
            icon &&
            !iconToLoad && (
              <BarCenter>
                <div className="sto-app-bar">
                  <img
                    className={`sto-app-bar-icon ${
                      !displayBarIcon ? 'sto-app-bar-icon--hidden' : ''
                    }`}
                    src={icon}
                    alt={`${slug}-icon`}
                  />
                </div>
              </BarCenter>
            )}
          <div className="sto-app">
            <Button
              icon="back"
              tag={Link}
              to={`${parent}`}
              onClick={this.unmountTrap}
              className="sto-app-back"
              label={t('app_page.back')}
              subtle
            />
            <Header
              app={app}
              namePrefix={namePrefix}
              name={appName}
              description={appShortDesc}
              parent={parent}
            />
            {app.screenshots &&
              !!app.screenshots.length && (
                <Gallery slug={slug} images={app.screenshots} />
              )}
            <Details
              app={app}
              description={appLongDesc}
              changes={appChanges}
              mobileApps={mobileApps}
              parent={parent}
            />
          </div>
        </div>
      </FocusTrap>
    )
  }
}

// translate is needed here for the lang props
export default translate()(withBreakpoints()(ApplicationPage))
