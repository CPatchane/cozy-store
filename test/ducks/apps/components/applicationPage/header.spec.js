'use strict'

/* eslint-env jest */

import React from 'react'
import { shallow } from 'enzyme'

import { tMock } from '../../../../jestLib/I18n'
import { Header } from 'ducks/apps/components/ApplicationPage/Header'

import mockApp from '../../_mockPhotosRegistryVersion'

window.location.assign = jest.fn()

describe('ApplicationPage header component', () => {
  it('should be rendered correctly provided app', () => {
    const component = shallow(
      <Header t={tMock} parent='/myapps' {...mockApp.manifest} />
    ).node
    expect(component).toMatchSnapshot()
  })

  it('should be rendered correctly provided app with broken icon and no editor', () => {
    const appProps = Object.assign({}, mockApp.manifest)
    appProps.icon = ''
    appProps.editor = ''
    const component = shallow(
      <Header t={tMock} parent='/myapps' {...appProps} />
    ).node
    expect(component).toMatchSnapshot()
  })

  it('should handle the click on open installed app button', () => {
    const appProps = Object.assign({}, mockApp.manifest)
    appProps.installed = true
    appProps.icon = '<svg></svg>'
    appProps.installedAppLink = '#'
    const component = shallow(
      <Header t={tMock} parent='/myapps' {...appProps} />
    )
    component.find('button[onClick]').simulate('click')
    expect(window.location.assign.mock.calls.length).toBe(1)
    expect(window.location.assign.mock.calls[0][0]).toBe('#')
  })
})
