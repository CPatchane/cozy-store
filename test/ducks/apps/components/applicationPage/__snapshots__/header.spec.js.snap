// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplicationPage header component should be rendered correctly provided app 1`] = `
<div
  className="sto-app-header"
>
  <div
    className="sto-app-header-icon"
  >
    <img
      alt="photos-icon"
      className="sto-app-icon"
      src="https://mockcozy.cc/registry/photos/3.0.0/icon.svg"
    />
  </div>
  <div
    className="sto-app-header-content"
  >
    <h2
      className="sto-app-header-title"
    >
      Cozy Photos
    </h2>
    <p
      className="sto-app-header-description"
    />
    <Link
      className="c-btn c-btn--regular"
      disabled={undefined}
      onClick={null}
      replace={false}
      to="//myapps/photos/manage"
    >
      <span>
        <Icon
          className="sto-app-icon--button"
          color="#FFFFFF"
          height="16px"
          icon="test-file-stub"
          spin={false}
          width="16px"
        />
         
        Install
      </span>
    </Link>
  </div>
</div>
`;

exports[`ApplicationPage header component should be rendered correctly provided app with broken icon and no editor 1`] = `
<div
  className="sto-app-header"
>
  <div
    className="sto-app-header-icon"
  >
    <Icon
      className="sto-app-icon--default blurry"
      color="#95999D"
      height="88px"
      icon="test-file-stub"
      spin={false}
      width="88px"
    />
  </div>
  <div
    className="sto-app-header-content"
  >
    <h2
      className="sto-app-header-title"
    >
      Photos
    </h2>
    <p
      className="sto-app-header-description"
    />
    <Link
      className="c-btn c-btn--regular"
      disabled={undefined}
      onClick={null}
      replace={false}
      to="//myapps/photos/manage"
    >
      <span>
        <Icon
          className="sto-app-icon--button"
          color="#FFFFFF"
          height="16px"
          icon="test-file-stub"
          spin={false}
          width="16px"
        />
         
        Install
      </span>
    </Link>
  </div>
</div>
`;

exports[`ApplicationPage header component should be rendered correctly provided konnector 1`] = `
<div
  className="sto-app-header"
>
  <div
    className="sto-app-header-icon"
  >
    <img
      alt="konnector-trinlane-icon"
      className="sto-app-icon"
      src="https://mockcozy.cc/registry/konnector-trinlane/0.1.0/icon.svg"
    />
  </div>
  <div
    className="sto-app-header-content"
  >
    <h2
      className="sto-app-header-title"
    >
      Trinlane
    </h2>
    <p
      className="sto-app-header-description"
    />
    <Link
      className="c-btn c-btn--regular"
      disabled={undefined}
      onClick={null}
      replace={false}
      to="//myapps/konnector-trinlane/manage"
    >
      <span>
        <Icon
          className="sto-app-icon--button"
          color="#FFFFFF"
          height="16px"
          icon="test-file-stub"
          spin={false}
          width="16px"
        />
         
        Install
      </span>
    </Link>
  </div>
</div>
`;
